### PR TITLE
Add fine grained metrics+logging for handling, processing, and grab l…

### PR DESCRIPTION
**Pull request by aaronbuchwald**
## Why this should be merged

This PR adds more fine grained tracking for the amount of time spent in message processing. Currently, message processing tracks the amount of time spent handling a message including the time spent grabbing the engine lock. This PR breaks down the handling time into:

- total handling time (includes grabbing the lock)
- acquire lock time (time it takes to grab the lock when processing a message)
- processing time (the time spent processing the message AFTER grabbing the lock)

Note: these metrics are observed both before and after bootstrapping, so new nodes that spend a long time bootstrapping may have very skewed metrics at the start.

## How this works

This PR tracks the new finer grained metrics and updates the warn log reported in https://github.com/ava-labs/avalanchego/issues/1286 to include all three, so that we have more information when this warn occurs.

## How this was tested

This PR is tested in CI and by running locally with the ANR to observe that the metrics show up as expected:

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP avalanche_P_handler_state_summary_frontier_count Total # of observations of time (in ns) of handling a state_summary_frontier
# TYPE avalanche_P_handler_state_summary_frontier_count counter
avalanche_P_handler_state_summary_frontier_count 0
# HELP avalanche_P_handler_state_summary_frontier_lock_count Total # of observations of time (in ns) of acquiring a lock to process a state_summary_frontier
# TYPE avalanche_P_handler_state_summary_frontier_lock_count counter
avalanche_P_handler_state_summary_frontier_lock_count 0
# HELP avalanche_P_handler_state_summary_frontier_lock_sum Sum of time (in ns) of acquiring a lock to process a state_summary_frontier
# TYPE avalanche_P_handler_state_summary_frontier_lock_sum gauge
avalanche_P_handler_state_summary_frontier_lock_sum 0
# HELP avalanche_P_handler_state_summary_frontier_processing_count Total # of observations of time (in ns) of processing a state_summary_frontier
# TYPE avalanche_P_handler_state_summary_frontier_processing_count counter
avalanche_P_handler_state_summary_frontier_processing_count 0
# HELP avalanche_P_handler_state_summary_frontier_processing_sum Sum of time (in ns) of processing a state_summary_frontier
# TYPE avalanche_P_handler_state_summary_frontier_processing_sum gauge
avalanche_P_handler_state_summary_frontier_processing_sum 0
# HELP avalanche_P_handler_state_summary_frontier_sum Sum of time (in ns) of handling a state_summary_frontier
# TYPE avalanche_P_handler_state_summary_frontier_sum gauge
avalanche_P_handler_state_summary_frontier_sum 0
100  484k    0  484k    0     0  23.7M      0 --:--:-- --:--:-- --:--:-- 36.3M
```

Note: I kept the name of the original metric for the total time spent handling the message the same, so that existing metrics are unchanged. This has the unfortunate side effect that the output now shows the count at the beginning and the sum at the end, with the other metrics in between, but this is only a problem when viewing the raw metrics with grep.

### CODE REVIEW

File: snow/networking/handler/handler.go [hunk 1]
```diff
@@ -423,6 +423,8 @@ func (h *handler) handleSyncMsg(ctx context.Context, msg Message) error {
 		op        = msg.Op()
 		body      = msg.Message()
 		startTime = h.clock.Time()
+		// Check if the chain is in normal operation at the start of message execution (may change during execution)
+		isNormalOp = h.ctx.State.Get().State == snow.NormalOp
```
Comment (aaronbuchwald) [inline snow/networking/handler/handler.go:427]: Per @StephenButtolph 's suggestion, we should care about whether the operation is in normal operation at the start of processing instead of the end.

If the state changes during the execution of the message, then we would still expect that this message may take a much longer time to process since the blockchain is still bootstrapping.

File: snow/networking/handler/handler.go [hunk 2]
```diff
@@ -435,23 +437,30 @@ func (h *handler) handleSyncMsg(ctx context.Context, msg Message) error {
...
+			messageHistograms = h.metrics.messages[op]
+			acquireLockTime   = lockAcquiredTime.Sub(startTime)
+			processingTime    = endTime.Sub(lockAcquiredTime)
+			handlingTime      = endTime.Sub(startTime)
 		)
 		h.resourceTracker.StopProcessing(nodeID, endTime)
-		histogram.Observe(float64(processingTime))
+		messageHistograms.handlingTime.Observe(float64(handlingTime))
+		messageHistograms.acquireLockTime.Observe(float64(acquireLockTime))
+		messageHistograms.processingTime.Observe(float64(processingTime))
```
Comment (aaronbuchwald) [inline snow/networking/handler/handler.go:454]: Open to naming changes - splitting this into handling time and processing time made sense to me

File: snow/networking/handler/metrics.go [hunk 3]
```diff
@@ -37,16 +43,33 @@ func newMetrics(namespace string, reg prometheus.Registerer) (*metrics, error) {
...
-			namespace,
-			opStr,
-			fmt.Sprintf("time (in ns) of processing a %s", opStr),
-			reg,
-			&errs,
-		)
+		messageProcessing := &messageProcessing{
+			handlingTime: metric.NewAveragerWithErrs(
+				namespace,
+				opStr,
```
Comment (aaronbuchwald) [inline snow/networking/handler/metrics.go:52]:
```suggestion
				opStr, // TODO: add a suffix to this metric in the future (this will change the existing metric)
```

File: snow/networking/handler/handler.go (hunk 2 above)
Comment (darioush) [reply snow/networking/handler/handler.go:454]: Should we leave processingTime as is (eg for continuity in metrics) and then add acquireLockTime and handlingTime (where processingTime = handlingTime + acquireLockTime)?

Comment (aaronbuchwald) [reply snow/networking/handler/handler.go:454]: Yup, that's what is happening here. I just renamed the variables in the code, since `handlingTime` seemed like a better name for the total time spent handling the message vs. actually processing it once the lock was acquired.

This means that `handlingTime = processingTime + acquireLockTime`, but the original metric, which we called `processingTime` still represents the same value.

**Review by darioush — APPROVED** [review]

**Review by darioush — APPROVED** [review]

File: snow/networking/handler/handler.go [hunk 4]
```diff
@@ -435,23 +437,27 @@ func (h *handler) handleSyncMsg(ctx context.Context, msg Message) error {
 	)
 	h.resourceTracker.StartProcessing(nodeID, startTime)
 	h.ctx.Lock.Lock()
```
Comment (joshua-kim) [inline snow/networking/handler/handler.go:439]: Wondering if it makes sense to have a metric for how long we're waiting to acquire this lock in addition to the metrics we're adding (how long this thread holding the lock).

File: snow/networking/handler/handler.go [hunk 5]
```diff
@@ -844,18 +853,30 @@ func (h *handler) handleChanMsg(msg message.InboundMessage) error {
 		zap.Any("message", body),
 	)
 	h.ctx.Lock.Lock()
+	lockAcquiredTime := h.clock.Time()
```
Comment (joshua-kim) [inline snow/networking/handler/handler.go:856]: Maybe we should call this `lockHeldStartTime`; `lockAcquireTime` sounds more like the time waiting to grab a lock

File: snow/networking/handler/handler.go [hunk 6]
```diff
@@ -844,18 +853,30 @@ func (h *handler) handleChanMsg(msg message.InboundMessage) error {
...
 	defer func() {
 		h.ctx.Lock.Unlock()
 
 		var (
-			endTime   = h.clock.Time()
-			histogram = h.metrics.messages[op]
+			endTime           = h.clock.Time()
+			messageHistograms = h.metrics.messages[op]
+			acquireLockTime   = lockAcquiredTime.Sub(startTime)
```
Comment (joshua-kim) [inline snow/networking/handler/handler.go:863]: maybe we can call this `lockHeldTime`

File: snow/networking/handler/metrics.go [hunk 7]
```diff
@@ -37,16 +42,26 @@ func newMetrics(namespace string, reg prometheus.Registerer) (*metrics, error) {
...
-		)
+		messageProcessing := &messageProcessing{
+			processingTime: metric.NewAveragerWithErrs(
+				namespace,
+				opStr,
+				fmt.Sprintf("time (in ns) of handling a %s", opStr),
+				reg,
+				&errs,
+			),
+			acquireLockTime: metric.NewAveragerWithErrs(
```
Comment (joshua-kim) [inline snow/networking/handler/metrics.go:56]: same comment from before

File: snow/networking/handler/metrics.go [hunk 8]
```diff
@@ -37,16 +42,26 @@ func newMetrics(namespace string, reg prometheus.Registerer) (*metrics, error) {
...
+				namespace,
+				fmt.Sprintf("%s_lock", opStr),
+				fmt.Sprintf("time (in ns) of acquiring a lock to process a %s", opStr),
```
Comment (joshua-kim) [inline snow/networking/handler/metrics.go:59]: maybe reword this to be `time (in ns) spent holding the lock to process...`?

**Comment by joshua-kim** [comment]
super minor nits, one question on if we should add a metric on acquire time in addition to the held time we're adding

File: snow/networking/handler/handler.go (hunk 5 above)
Comment (StephenButtolph) [reply snow/networking/handler/handler.go:856]: I think because it is `Acquired` (past tense) it's pretty clear.

File: snow/networking/handler/handler.go (hunk 4 above)
Comment (aaronbuchwald) [reply snow/networking/handler/handler.go:439]: As per @darioush 's comment, I think it's better to avoid adding a metric that we can derive from the other ones

**Comment by aaronbuchwald** [comment]
Good call @joshua-kim

**Review by StephenButtolph — APPROVED** [review]

**Review by joshua-kim — APPROVED** [review]
